File: internal/params/varmap.go
package params

import (
	"fmt"
	"strings"
)

// The var map is the firmware's ordered list of runtime variables (device.cpp
// InitVarMap). Input fields (output[n].input, condition[n].input, etc.) reference
// a variable by its index here. We expose stable names so configs read clearly;
// a raw index is also always accepted.
//
// The order below is InitVarMap's order exactly — system vars, digital inputs,
// digital OUTPUTS, analog inputs, CAN inputs, virtual inputs, outputs, flashers,
// conditions, counters, wiper, keypads — not the order the VAR_MAP_SIZE macro
// happens to sum its terms in (those disagree between boards and the macro is
// only a sum, so only InitVarMap is authoritative).
//
// Names are 1-based to match the UI.

// VarName returns the human name for a var-map index ("" if out of range).
func (r *Registry) VarName(i uint16) string {
	if int(i) < len(r.varNames) {
		return r.varNames[i]
	}
	return ""
}

// VarIndex resolves a var-map name (case-insensitive) to its index.
func (r *Registry) VarIndex(name string) (uint16, bool) {
	i, ok := r.varByName[strings.ToLower(strings.TrimSpace(name))]
	return i, ok
}

// VarMapSize is the number of var-map slots for this board.
func (r *Registry) VarMapSize() int { return len(r.varNames) }

// VarNames returns every var-map name in index order.
func (r *Registry) VarNames() []string { return r.varNames }

// VarName resolves against the default board.
func VarName(i uint16) string { return defaultReg.VarName(i) }

// VarIndex resolves against the default board.
func VarIndex(name string) (uint16, bool) { return defaultReg.VarIndex(name) }

// VarMapSize reports the default board's var-map size.
func VarMapSize() int { return defaultReg.VarMapSize() }

func (r *Registry) vadd(name string) { r.varNames = append(r.varNames, name) }

func (r *Registry) buildVarMap() {
	bd := r.board

	// System vars. AlwaysFalse/AlwaysTrue/State are unconditional; BoardTemp
	// and BattVolt exist only where HAS_EXT_TEMP_SENSOR / HAS_BATT_VOLT_SENSE,
	// which is what SysVars (3 or 5) encodes.
	r.vadd("AlwaysFalse")
	r.vadd("AlwaysTrue")
	r.vadd("State")
	if bd.SysVars >= 4 {
		r.vadd("BoardTemp")
	}
	if bd.SysVars >= 5 {
		r.vadd("BattVolt")
	}
	// Digital inputs (1 var each).
	for n := 1; n <= bd.DigInputs; n++ {
		r.vadd(fmt.Sprintf("DigIn%d", n))
	}
	// Digital outputs (1 var each) — firmware InitVarMap order: digOut before
	// analog and CAN inputs.
	for n := 1; n <= bd.DigOutputs; n++ {
		r.vadd(fmt.Sprintf("DigOut%d", n))
	}
	// Analog inputs (4 vars each: Value, milliVolts, RotaryPos, SwitchVal).
	for n := 1; n <= bd.AnalogInputs; n++ {
		r.vadd(fmt.Sprintf("Analog%dVal", n))
		r.vadd(fmt.Sprintf("Analog%dmV", n))
		r.vadd(fmt.Sprintf("Analog%dRotary", n))
		r.vadd(fmt.Sprintf("Analog%dSwitch", n))
	}
	// CAN inputs (2 vars each: the boolean Output of the operator comparison,
	// then the decoded numeric Value).
	for n := 1; n <= bd.CanInputs; n++ {
		r.vadd(fmt.Sprintf("CanIn%dOut", n))
		r.vadd(fmt.Sprintf("CanIn%dVal", n))
	}
	// Virtual inputs (1 var each).
	for n := 1; n <= bd.VirtInputs; n++ {
		r.vadd(fmt.Sprintf("VirtIn%d", n))
	}
	// Outputs (4 vars each: Active, Current, Overcurrent, Fault).
	for n := 1; n <= bd.Outputs; n++ {
		r.vadd(fmt.Sprintf("Out%dActive", n))
		r.vadd(fmt.Sprintf("Out%dCurrent", n))
		r.vadd(fmt.Sprintf("Out%dOvercurrent", n))
		r.vadd(fmt.Sprintf("Out%dFault", n))
	}
	// Flashers (1 var each).
	for n := 1; n <= bd.Flashers; n++ {
		r.vadd(fmt.Sprintf("Flasher%d", n))
	}
	// Conditions (1 var each).
	for n := 1; n <= bd.Conditions; n++ {
		r.vadd(fmt.Sprintf("Cond%d", n))
	}
	// Counters (1 var each).
	for n := 1; n <= bd.Counters; n++ {
		r.vadd(fmt.Sprintf("Counter%d", n))
	}
	// Wiper outputs (VAR_MAP_WIPER_VARS = 6).
	if bd.HasWipers {
		r.vadd("WiperSlowOut")
		r.vadd("WiperFastOut")
		r.vadd("WiperParkOut")
		r.vadd("WiperInterOut")
		r.vadd("WiperWashOut")
		r.vadd("WiperSwipeOut")
	}
	// Keypads (each: buttons, dials, analog).
	for k := 1; k <= bd.Keypads; k++ {
		for b := 1; b <= bd.KeypadButtons; b++ {
			r.vadd(fmt.Sprintf("Keypad%dButton%d", k, b))
		}
		for d := 1; d <= bd.KeypadDials; d++ {
			r.vadd(fmt.Sprintf("Keypad%dDial%d", k, d))
		}
		for a := 1; a <= bd.KeypadAnalogs; a++ {
			r.vadd(fmt.Sprintf("Keypad%dAnalog%d", k, a))
		}
	}

	r.varByName = make(map[string]uint16, len(r.varNames))
	for i, n := range r.varNames {
		r.varByName[strings.ToLower(n)] = uint16(i)
	}
}

File: fwmodel/model/varmap.py
"""Board definitions and var-map layout.

Transcribed from dingoFW:
  - boards/<variant>/port.h            (the NUM_* / VAR_MAP_* counts)
  - core/device.cpp  InitVarMap()      (the *order* the slots are laid out in)

InitVarMap() order is the authority, NOT the VAR_MAP_SIZE macro (the macro is a
sum, so its term order is meaningless; pt-dpdm4_1 and dingopdmmax_v1 list canIn
before analogIn in the macro but InitVarMap always does analog first).

Order, per core/device.cpp:
    sys vars, digIn, digOut, analogIn(x4), canIn(x2), virtIn,
    outputs(x4), flashers, conditions, counters, wiper(6), keypads

Sys vars are conditional:
    [0] AlwaysFalse, [1] AlwaysTrue, [2] State,
    [3] BoardTemp   (only if HAS_EXT_TEMP_SENSOR),
    [4] BattVolt    (only if HAS_BATT_VOLT_SENSE)
so VAR_MAP_SYS_VARS is 5 on the PDMs and 3 on canboard_v2.
"""

from dataclasses import dataclass, field
from typing import Dict, List


@dataclass
class Board:
    name: str
    pdm_type: int          # PDM_TYPE, or -1 where the board has none (canboard)
    sys_vars: int
    num_dig_inputs: int
    num_dig_outputs: int
    num_analog_inputs: int
    num_can_inputs: int
    num_can_outputs: int
    num_virt_inputs: int
    num_outputs: int
    num_flashers: int
    num_conditions: int
    num_counters: int
    wiper_vars: int
    num_keypads: int
    keypad_buttons: int = 0
    keypad_dials: int = 0
    keypad_analogs: int = 0

    names: List[str] = field(default_factory=list, repr=False)
    by_name: Dict[str, int] = field(default_factory=dict, repr=False)
    base: Dict[str, int] = field(default_factory=dict, repr=False)

    def __post_init__(self) -> None:
        n: List[str] = []
        base: Dict[str, int] = {}

        def block(key: str, items):
            base[key] = len(n)
            n.extend(items)

        sysnames = ["AlwaysFalse", "AlwaysTrue", "State", "BoardTemp", "BattVolt"]
        block("sys", sysnames[: self.sys_vars])
        block("digIn", [f"DigIn{i}" for i in range(1, self.num_dig_inputs + 1)])
        block("digOut", [f"DigOut{i}" for i in range(1, self.num_dig_outputs + 1)])
        block(
            "analogIn",
            [
                s
                for i in range(1, self.num_analog_inputs + 1)
                for s in (
                    f"Analog{i}Val",
                    f"Analog{i}mV",
                    f"Analog{i}Rotary",
                    f"Analog{i}Switch",
                )
            ],
        )
        block(
            "canIn",
            [
                s
                for i in range(1, self.num_can_inputs + 1)
                for s in (f"CanIn{i}Out", f"CanIn{i}Val")
            ],
        )
        block("virtIn", [f"VirtIn{i}" for i in range(1, self.num_virt_inputs + 1)])
        block(
            "output",
            [
                s
                for i in range(1, self.num_outputs + 1)
                for s in (
                    f"Out{i}Active",
                    f"Out{i}Current",
                    f"Out{i}Overcurrent",
                    f"Out{i}Fault",
                )
            ],
        )
        block("flasher", [f"Flasher{i}" for i in range(1, self.num_flashers + 1)])
        block("condition", [f"Cond{i}" for i in range(1, self.num_conditions + 1)])
        block("counter", [f"Counter{i}" for i in range(1, self.num_counters + 1)])
        block(
            "wiper",
            [
                "WiperSlowOut",
                "WiperFastOut",
                "WiperParkOut",
                "WiperInterOut",
                "WiperWashOut",
                "WiperSwipeOut",
            ][: self.wiper_vars],
        )
        kp = []
        for k in range(1, self.num_keypads + 1):
            kp += [f"Keypad{k}Button{b}" for b in range(1, self.keypad_buttons + 1)]
            kp += [f"Keypad{k}Dial{d}" for d in range(1, self.keypad_dials + 1)]
            kp += [f"Keypad{k}Analog{a}" for a in range(1, self.keypad_analogs + 1)]
        block("keypad", kp)

        self.names = n
        self.base = base
        self.by_name = {s.lower(): i for i, s in enumerate(n)}

    @property
    def size(self) -> int:
        return len(self.names)

    def index(self, name: str) -> int:
        return self.by_name[name.strip().lower()]

    def label(self, idx: int) -> str:
        if 0 <= idx < len(self.names):
            return self.names[idx]
        return f"<oob {idx}>"


_PDM_KEYPAD = dict(num_keypads=2, keypad_buttons=20, keypad_dials=2, keypad_analogs=4)

BOARDS: Dict[str, Board] = {}


def _reg(b: Board) -> Board:
    BOARDS[b.name] = b
    return b


# dingoFW/boards/dingopdm_v7/port.h   VAR_MAP_SIZE = 217
DINGOPDM_V7 = _reg(
    Board(
        name="dingopdm_v7",
        pdm_type=0,
        sys_vars=5,
        num_dig_inputs=2,
        num_dig_outputs=0,
        num_analog_inputs=0,
        num_can_inputs=32,
        num_can_outputs=32,
        num_virt_inputs=16,
        num_outputs=8,
        num_flashers=4,
        num_conditions=32,
        num_counters=4,
        wiper_vars=6,
        **_PDM_KEYPAD,
    )
)

# dingoFW/boards/dingopdmmax_v1/port.h   VAR_MAP_SIZE = 201
DINGOPDMMAX_V1 = _reg(
    Board(
        name="dingopdmmax_v1",
        pdm_type=1,
        sys_vars=5,
        num_dig_inputs=2,
        num_dig_outputs=0,
        num_analog_inputs=0,
        num_can_inputs=32,
        num_can_outputs=32,
        num_virt_inputs=16,
        num_outputs=4,
        num_flashers=4,
        num_conditions=32,
        num_counters=4,
        wiper_vars=6,
        **_PDM_KEYPAD,
    )
)

# dingoFW/boards/pt-dpdm4_1/port.h   VAR_MAP_SIZE = 209
PT_DPDM4_1 = _reg(
    Board(
        name="pt-dpdm4_1",
        pdm_type=2,
        sys_vars=5,
        num_dig_inputs=2,
        num_dig_outputs=0,
        num_analog_inputs=2,
        num_can_inputs=32,
        num_can_outputs=32,
        num_virt_inputs=16,
        num_outputs=4,
        num_flashers=4,
        num_conditions=32,
        num_counters=4,
        wiper_vars=6,
        **_PDM_KEYPAD,
    )
)

# dingoFW/boards/canboard_v2/port.h   VAR_MAP_SIZE = 75
CANBOARD_V2 = _reg(
    Board(
        name="canboard_v2",
        pdm_type=-1,
        sys_vars=3,
        num_dig_inputs=8,
        num_dig_outputs=4,
        num_analog_inputs=5,
        num_can_inputs=8,
        num_can_outputs=8,
        num_virt_inputs=8,
        num_outputs=0,
        num_flashers=4,
        num_conditions=8,
        num_counters=4,
        wiper_vars=0,
        num_keypads=0,
    )
)

BY_PDM_TYPE = {b.pdm_type: b for b in BOARDS.values() if b.pdm_type >= 0}

# Pins the firmware exposes as constants; these are *ground truth anchors* used by
# selftest.py. Each entry is (board, var name, expected index) and was read off
# port.h + InitVarMap by hand.
ANCHORS = [
    ("dingopdm_v7", "AlwaysFalse", 0),
    ("dingopdm_v7", "AlwaysTrue", 1),
    ("dingopdm_v7", "DigIn1", 5),
    ("dingopdm_v7", "CanIn1Out", 7),
    ("dingopdm_v7", "VirtIn1", 71),
    ("dingopdm_v7", "Out1Active", 87),
    ("dingopdm_v7", "Flasher1", 119),
    ("dingopdm_v7", "Cond1", 123),
    ("dingopdm_v7", "Counter1", 155),
    ("dingopdm_v7", "WiperSlowOut", 159),
    ("dingopdmmax_v1", "VirtIn1", 71),
    ("dingopdmmax_v1", "Cond1", 107),
    ("dingopdmmax_v1", "Counter1", 139),
    ("pt-dpdm4_1", "VirtIn1", 79),
    ("pt-dpdm4_1", "Cond1", 115),
    ("pt-dpdm4_1", "Counter1", 147),
    ("canboard_v2", "VirtIn1", 51),
    ("canboard_v2", "Cond1", 63),
    ("canboard_v2", "Counter1", 71),
]

SIZES = {
    "dingopdm_v7": 217,
    "dingopdmmax_v1": 201,
    "pt-dpdm4_1": 209,
    "canboard_v2": 75,
}
